File: js/nrpc/nrpc.js
/*
 * @Author: FengyunSky lifengsky123@gmail.com
 * @Description: 对接核心 rpc 层实现远程过程调用，实现核心的rpc主要调用基础API
 */

const Assert = require('assert');
const String = require('string');
const Is = require('is');
const Util = require('util');
const Nats = require('nats');
const NatsClient = require('../nats/natsclient/nats_client');
const Proto = require('./nrpc.pb');
const Response = Proto.nrpc.Response;
const Request = Proto.nrpc.Request;
const ResponseResult = Proto.nrpc.ResponseResult;

const EncodeType = {
    String: 'string',
    Json: 'json',
    Protobuf: 'protobuf'
}

const EncodeType1 = {
    Request: 0, //请求
    Response: 1,//响应
}

const ExpectedReplySujectHdr = "Nats-Expented-Reply-Subject";
const ExpectedStreamHdr      = "Nats-Expected-Stream";
const MsgIdHdr               = "Nats-Msg-Id";
const ExpectedLastSeqHdr     = "Nats-Expected-Last-Sequence";
const ExpectedLastSubjSeqHdr = "Nats-Expected-Last-Subject-Sequence";
const ExpectedLastMsgIdHdr   = "Nats-Expected-Last-Msg-Id";
const MsgRollup              = "Nats-Rollup";

class NRPC {
    static StringCode = Nats.StringCodec();
    static JsonCode = Nats.JSONCodec();

    constructor(name, subject, client, encodeType = EncodeType.Json) {
        Assert(Is.string(name) && name != '', 'nrpc constructor failed, invalid name');
        Assert(Is.string(subject) && subject != '', 'nrpc constructor failed, invalid subject');
        Assert(client instanceof NatsClient.NatsClient, 'nrpc constructor failed, invalid client');
        Assert(encodeType === EncodeType.String || encodeType === EncodeType.Json || 
                encodeType === EncodeType.Protobuf, 'nrpc constructor failed, invalid encoding');

        this.name = name;
        this.subject = subject;
        this.client = client;
        this.encodeType = encodeType;
    }

    /**
     * @name: bindStream
     * @msg: 绑定stream
     * @param {StreamConfig} streamCfg
     * @return {Promise(StreamInfo)}
     * @abstract 添加失败可能是已经添加忽略错误
     */    
    async bindStream(streamCfg) {
        Assert(Is.object(streamCfg), 'bindStream failed, invalid streamCfg, not object');
        Assert(Is.array(streamCfg.subjects), 'bindStream failed, invalid streamCfg for subjects, not array');

        let isvalid = false;
        streamCfg.subjects.forEach(subject => {
            if (String(subject).startsWith(this.subject)) {
                isvalid = true;
                return;
            }
        });

        if (!isvalid) {
            Assert(false, Util.format('bindStream failed, invalid streamCfg for subjects, not start with %s', this.subject));
        }

        try {
            const streamInfo = await this.client.addStream(streamCfg);
            return Promise.resolve(streamInfo);
        } catch (err) {
            console.log(Util.format('bindStream failed, addStream err:%s', err.message));
            return Promise.resolve();
        }
    }

    /**
     * @name: bindConsumer
     * @msg: 绑定消费者
     * @param {string} stream
     * @param {ConsumerConfig} consumerCfg
     * @return {Promise(ConsumerInfo)}
     */    
    async bindConsumer(stream, consumerCfg) {
        Assert(Is.string(stream) && stream != '', 'bindConsumer failed, invalid stream');
        return this.client.addConsumer(stream, consumerCfg);
    }

    /**
     * @name: checkStreamWithSubject
     * @msg: 检查是否主题匹配指定的stream
     * @param {string} stream
     * @param {string} subject
     * @return {Promise}
     */    
    async checkStreamWithSubject(stream, subject) {
        Assert(Is.string(stream), 'checkStreamWithSubject failed, invalid stream');
        Assert(Is.string(subject) && subject != '', 'checkStreamWithSubject failed, invalid subject');

        try {
            const name = await this.client.streamNameBySubject(subject);
            if (name !== stream) {
                return Promise.reject(new Error(Util.format('checkStreamWithSubject failed, not found stream:%s for subject:%s',
                                    stream, subject)));
            }
        } catch (err) {
            return Promise.reject(new Error(Util.format('checkStreamWithSubject failed, streamNameBySubject failed, err:%s',
                                    err.message)));
        }
        
        return Promise.resolve();
    }

    /**
     * @name: checkConsumerAndStreamWithSubject
     * @msg: 检查stream与消费者及主题是否匹配
     * @param {string} stream
     * @param {string} consumer
     * @param {string} subject
     * @return {Promise}
     */    
    async checkConsumerAndStreamWithSubject(stream, consumer, subject) {
        Assert(Is.string(stream), 'checkConsumerAndStreamWithSubject failed, invalid stream');
        Assert(Is.string(consumer), 'checkConsumerAndStreamWithSubject failed, invalid consumer');
        Assert(Is.string(subject) && subject != '', 'checkConsumerAndStreamWithSubject failed, invalid subject');

        Assert(stream == '' && consumer != '', 'checkConsumerAndStreamWithSubject failed, invalid empty stream for consumer');

        try {
            const name = await this.client.streamNameBySubject(subject);
            if (name != stream) {
                return Promise.reject(new Error(Util.format('checkConsumerAndStreamWithSubject failed, the stream(%s) for subject(%s) not match the input stream(%s)',
                        stream, subject, name)));
            }
        } catch (err) {
            if (stream != '') {
                return Promise.reject(new Error(Util.format('checkConsumerAndStreamWithSubject failed, not found the stream(%s) for subject(%s)',
                        stream, subject)));
            }
            
            return Promise.resolve();
        }

        try {
            const consumerInfo = await this.client.consumerInfoByStream(stream, consumer);
            if (!String(consumerInfo.config.filter_subject).startsWith(this.subject)) {
                return Promise.reject(
                    new Error(Util.format('checkConsumerAndStreamWithSubject failed, the filter subjectes(%s) of consumer(%s), not match the subject:%s"',
                                consumerInfo.config.filter_subject, consumer, subject)));
            }
        } catch (err) {
            return Promise.reject( 
                new Error(Util.format('checkConsumerAndStreamWithSubject failed, the consumer(%s) not found for stream(%s), ConsumerInfoByStream err:%s',
                consumer, stream, err.message)));
        }
    }

    /**
     * @name: checkSubject
     * @msg: 校验主题
     * @param {string} subject
     * @return {void}
     */    
    checkSubject(subject) {
        Assert(Is.string(subject) && subject !== '' && 
        (String(subject).contains('.*') || String(subject).contains('.>')),
        Util.format('checkSubject failed, invalid subject(%s), must not empty or .* or .>', subject));
    }

    /**
     * @name: encode
     * @msg: 序列化编码
     * @param {EncodeType} encodeType
     * @param {string|Uint8Array} data
     * @param {EncodeType1} type
     * @return {Uint8Array}
     */    
    static encode(encodeType, data, type=EncodeType1.Response) {
        switch(encodeType) {
        case (EncodeType.String):
            return NRPC.StringCode.encode(data);
        case (EncodeType.Json):
            return NRPC.JsonCode.encode(data);
        case (EncodeType.Protobuf):
            if (type == EncodeType1.Request) {
                return Request.encode(data).finish();
            }
            return Response.encode(data).finish();
        default:
            Assert(false, Util.format('encode failed, invalid encode:%s', encodeType));
        }
    }

    /**
     * @name: decode
     * @msg: 序列化解码
     * @param {EncodeType} encodeType
     * @param {Uint8Array} data
     * @param {EncodeType1} type
     * @return {*}
     */    
    static decode(encodeType, data, type=EncodeType1.Response) {
        switch(encodeType) {
        case (EncodeType.String):
            return NRPC.StringCode.decode(data);
        case (EncodeType.Json):
            return NRPC.JsonCode.decode(data);
        case (EncodeType.Protobuf):
            if (type == EncodeType1.Request) {
                return Request.decode(data);
            }
            return Response.decode(data);
        default:
            Assert(false, Util.format('decode failed, invalid encode:%s', encodeType));
        }
    }

    /**
     * @name: getCallerFuncName
     * @msg: 获取调用者函数名称
     * @return {string}
     */    
    static getCallerFuncName() {
        const stackArray = (new Error().stack).split('\n');
        const funMathcer = /([^(]+)@|at ([^(]+) \(/;
        const regexResult = funMathcer.exec(stackArray[3]);
        const callerName = regexResult[1] || regexResult[2];
        return callerName.split('.')[1];
    }

    /**
     * @name: getCurrentFuncName
     * @msg: 获取当前函数的名称
     * @return {string}
     */    
    static getCurrentFuncName() {
        const stackArray = (new Error().stack).split('\n');
        const funMathcer = /([^(]+)@|at ([^(]+) \(/;
        const regexResult = funMathcer.exec(stackArray[2]);
        const callerName = regexResult[1] || regexResult[2];
        return callerName.split('.')[1];
    }

    /**
     * @name: getFunArgsName
     * @msg: 获取函数参数名称数组
     * @param {function} func 函数
     * @return {Array} 函数参数名称数组
     */    
    static getFunArgsName(func) {
        Assert(Is.function(func), 'getFunArgsName failed, invalid func');
        // 先用正则匹配,取得符合参数模式的字符串.
        // 第一个分组是这个:  ([^)]*) 非右括号的任意字符
        var args = func.toString().match(/\(\s*([^)]+?)\s*\)/)[1];
        // 用逗号来分隔参数(arguments string).
        return args.split(",").map(function(arg) {
            // 去除注释(inline comments)以及空格
            return arg.replace(/\/\*.*\*\//, "").trim();
        }).filter(function(arg) {
            // 确保没有 undefined.
            return arg;
        });
      }

    /**
     * @name: response
     * @msg: 创建响应对象
     * @param {ResponseResult} result 响应结果
     * @param {string} message 响应结果描述
     * @param {array} args 响应结果参数列表
     * @return {Response}
     */    
    static response(result, message='', ...args) {
        Assert(Is.string(message), 'response failed, invalid message');
        //组装数据对象
        let data = {};
        for(let index = 0; index < arguments.length-2; index++) {
            data[index] = arguments[index+2];
        }

        let response = Response.create();
        response.result = result;
        response.message = message;
        response.data = Buffer.from(JSON.stringify(data));
        return response;
    }

    /**
     * @name: request
     * @msg: 创建reqeust请求对象
     * @param {Array[string]} names 调用函数的参数名称列表
     * @param {array} args 调用函数的参数列表
     * @return {Request}
     */    
    static request(names, ...args) {
        Assert(Is.array(names), 'request failed, invalid names, not the array');
        //组装数据对象
        let data = {};
        for(let index = 0; index < arguments.length-1; index++) {
            data[names[index]] = arguments[index+1];
        }

        let request = Request.create();
        request.name = NRPC.getCallerFuncName();
        request.data = Buffer.from(JSON.stringify(data));
        return request;
    }
      
}

module.exports = {
    NRPC,
    Request,
    Response,
    ResponseResult,
    EncodeType,
    EncodeType1,
    ExpectedReplySujectHdr,
    ExpectedStreamHdr,
}

File: go/nrpc/nrpc.go
/*
 * @Author: FengyunSky lifengsky123@gmail.com
 * @Description: 对接核心 rpc 层实现远程过程调用，实现核心的rpc主要调用基础API
 */

package nrpc

import (
	"errors"
	"log"
	"runtime"
	"strings"
	"time"

	"github.com/nats-io/nats.go"
	jsonpb "google.golang.org/protobuf/encoding/protojson"
	"google.golang.org/protobuf/proto"
)

const (
	LogTag = "nrpc"
)

const (
	EncodeJson     = "json"
	EncodeProtobuf = "protobuf"
)

const (
	ExpectedReplySujectHdr = "Nats-Expented-Reply-Subject"
)

type Logger interface {
	Verbosef(format string, args ...interface{})
	Infof(format string, args ...interface{})
	Warnf(format string, args ...interface{})
	Errorf(format string, args ...interface{})
	Panicf(format string, args ...interface{})
	Fatalf(format string, args ...interface{})
}

type NatsClient interface {
	AddStream(cfg *nats.StreamConfig) (*nats.StreamInfo, error)
	StreamNameBySubject(string, ...nats.JSOpt) (string, error)
	Publish(subj string, data []byte, opts ...nats.PubOpt) error
	PublishMsg(m *nats.Msg, opts ...nats.PubOpt) error
	Request(subj string, data []byte, timeout time.Duration) (*nats.Msg, error)
	PublishRequest(subj, reply string, data []byte) error

	AddConsumer(stream string, cfg *nats.ConsumerConfig, opts ...nats.JSOpt) (*nats.ConsumerInfo, error)
	ConsumerInfoByStream(stream, consumer string) (*nats.ConsumerInfo, error)
	ChanSubscribe(subj string, ch chan *nats.Msg, opts ...nats.SubOpt) (*nats.Subscription, error)
	QueueSubscribe(subj, queue string, cb nats.MsgHandler, opts ...nats.SubOpt) (*nats.Subscription, error)
	Subscribe(subj string, handler nats.MsgHandler, opts ...nats.SubOpt) (*nats.Subscription, error)
	SubscribeSync(subj string, opts ...nats.SubOpt) (*nats.Subscription, error)
	Unsubscribe(*nats.Subscription) error
}

type NRPC struct {
	name     string
	subject  string
	encoding string
	nc       NatsClient
	logger   Logger
}

func NewNPRC(name, subject, encoding string, nc NatsClient, logger Logger) *NRPC {
	return &NRPC{
		name:     name,
		subject:  subject,
		encoding: encoding,
		nc:       nc,
		logger:   logger,
	}
}

/**
 * @name: BindStream
 * @msg: 绑定流
 * @param {*nats.StreamConfig} cfg
 * @return {*}
 */
func (nrpc *NRPC) BindStream(cfg *nats.StreamConfig) (*nats.StreamInfo, error) {
	//绑定主题必须包含服务绑定的主题
	validSub := false
	for _, sub := range cfg.Subjects {
		if strings.HasPrefix(sub, nrpc.subject) {
			validSub = true
			break
		}
	}
	if !validSub {
		nrpc.logger.Panicf(LogTag, "BindStream failed, invalid subject")
	}
	return nrpc.nc.AddStream(cfg)
}

/**
 * @name: BindConsumer
 * @msg: 绑定消费者
 * @param {string} stream
 * @param {*nats.ConsumerConfig} cfg
 * @return {*}
 */
func (nrpc *NRPC) BindConsumer(stream string, cfg *nats.ConsumerConfig) (*nats.ConsumerInfo, error) {
	//消费者过滤的主题必须包含服务主题
	if !strings.HasPrefix(cfg.FilterSubject, nrpc.subject) {
		nrpc.logger.Panicf(LogTag, "BindConsumer failed, invalid subject:%s for filetersubject;%s", nrpc.subject, cfg.FilterSubject)
	}

	consumer, err := nrpc.nc.AddConsumer(stream, cfg)
	if err != nil {
		nrpc.logger.Warnf(LogTag, "BindConsumer failed, AddConsumer err:%v", err)
	}

	return consumer, err
}

/**
 * @name: Unmarshal
 * @msg: 解码
 * @param {string} encoding
 * @param {[]byte} data
 * @param {proto.Message} msg
 * @return {*}
 */
func Unmarshal(encoding string, data []byte, msg proto.Message) error {
	switch encoding {
	case EncodeProtobuf:
		return proto.Unmarshal(data, msg)
	case EncodeJson:
		return jsonpb.Unmarshal(data, msg)
	default:
		return errors.New("Invalid encoding: " + encoding)
	}
}

/**
 * @name: UnmarshalResponse
 * @msg: 解码响应
 * @param {string} encoding
 * @param {[]byte} data
 * @param {proto.Message} msg
 * @return {*}
 */
func UnmarshalResponse(encoding string, data []byte, msg proto.Message) error {
	switch encoding {
	case EncodeProtobuf:
		return proto.Unmarshal(data, msg)
	case EncodeJson:
		return jsonpb.Unmarshal(data, msg)
	default:
		return errors.New("Invalid encoding: " + encoding)
	}
}

/**
 * @name: Marshal
 * @msg: 编码
 * @param {string} encoding
 * @param {proto.Message} msg
 * @return {*}
 */
func Marshal(encoding string, msg proto.Message) ([]byte, error) {
	switch encoding {
	case EncodeProtobuf:
		return proto.Marshal(msg)
	case EncodeJson:
		return jsonpb.Marshal(msg)
	default:
		return nil, errors.New("Invalid encoding: " + encoding)
	}
}

/**
 * @name: ParseRequest
 * @msg: 解析请求对象
 * @param {string} encoding 编码方式
 * @param {[]byte} data 解析数据
 * @param {proto.Message} req
 * @return {*}
 */
func ParseRequest(encoding string, data []byte, req proto.Message) {
	err := Unmarshal(encoding, data, req)
	if err != nil {
		log.Panicf(" parse request failed, err:%v \n", err)
	}
}

/**
 * @name: NewResponse
 * @msg: 创建响应对象
 * @param {ResponseResult} result
 * @param {int32} count
 * @param {[]byte} data
 * @return {*}
 */
func NewResponse(result ResponseResult, message string, data []byte) *Response {
	return &Response{
		Result:  result,
		Message: message,
		Data:    data,
	}
}

/**
 * @name: CheckSubject
 * @msg: 校验主题
 * @param {string} subject
 * @return {*}
 */
func CheckSubject(subject string) {
	if subject == "" &&
		(!strings.Contains(subject, ".*") || !strings.Contains(subject, ".>")) {
		log.Panicf(LogTag, "CheckSubject failed, invalid subject, must not empty or .* or .>")
	}
}

/**
 * @name: CheckEncode
 * @msg: 校验编码
 * @param {string} encoding
 * @return {*}
 */
func CheckEncode(encoding string) {
	if encoding != EncodeProtobuf &&
		encoding != EncodeJson {
		log.Panicf(LogTag, "CheckEncode failed, invalid encoding, must be protobuf or json or string")
	}
}

/**
 * @name: CheckStreamWithSubject
 * @msg: 检查是否主题匹配指定的stream
 * @param {*} streamName
 * @param {string} subject
 * @return {*}
 */
func (nrpc *NRPC) CheckStreamWithSubject(streamName, subject string) {
	if subject == "" && streamName == "" {
		nrpc.logger.Panicf(LogTagServer, "CheckStreamWithSubject failed, invalid subject or stream, is empty")
	}

	stream, err := nrpc.nc.StreamNameBySubject(subject)
	if err != nil {
		if streamName != "" {
			nrpc.logger.Errorf(LogTagServer, "CheckStreamWithSubject failed, not found the stream(%s) for subject(%s)",
				streamName, subject)
		}

		return
	}
	if stream != streamName {
		nrpc.logger.Errorf(LogTagServer, "CheckStreamWithSubject failed, the stream(%s) for subject(%s) not match the input stream(%s)",
			stream, subject, streamName)
	}
}

/**
 * @name: CheckConsumerAndStreamWithSubject
 * @msg: 检查stream与消费者及主题是否匹配
 * @param {*} streamName
 * @param {*} consumerName
 * @param {string} subject
 * @return {*}
 */
func (nrpc *NRPC) CheckConsumerAndStreamWithSubject(streamName, consumerName, subject string) {
	if subject == "" {
		nrpc.logger.Panicf(LogTagServer, "CheckConsumerAndStreamWithSubject failed, invalid subject, is empty")
	}

	if streamName == "" && consumerName != "" {
		nrpc.logger.Panicf(LogTagServer, "CheckConsumerAndStreamWithSubject failed, invalid empty stream for consumer")
	}

	stream, err := nrpc.nc.StreamNameBySubject(subject)
	if err != nil {
		if streamName != "" {
			nrpc.logger.Panicf(LogTagServer, "CheckConsumerAndStreamWithSubject failed, not found the stream(%s) for subject(%s)",
				streamName, subject)
		}

		return
	}
	if stream != streamName {
		nrpc.logger.Panicf(LogTagServer, "CheckConsumerAndStreamWithSubject failed, the stream(%s) for subject(%s) not match the input stream(%s)",
			stream, subject, streamName)
	}

	consumer, err := nrpc.nc.ConsumerInfoByStream(streamName, consumerName)
	if err != nil {
		nrpc.logger.Panicf(LogTagServer,
			"CheckConsumerAndStreamWithSubject failed, the consumer(%s) not found for stream(%s), ConsumerInfoByStream err:%v",
			consumerName, streamName, err)
	}

	if !strings.HasPrefix(consumer.Config.FilterSubject, nrpc.subject) {
		nrpc.logger.Panicf(LogTagServer,
			"CheckConsumerAndStreamWithSubject failed, the filter subjectes(%s) of consumer(%s), not match the subject:%s",
			consumer.Config.FilterSubject, consumerName, subject)
	}
}

/**
 * @name: GetCurrentFuncName
 * @msg: 获取正在运行的函数名
 * @return {*}
 */
func GetCurrentFuncName() string {
	pc, _, _, _ := runtime.Caller(1)
	name := runtime.FuncForPC(pc).Name()
	names := strings.Split(name, ".")
	return names[len(names)-1]
}

/**
 * @name: GetCallerFuncName
 * @msg: 获取调用者函数名称
 * @return {*}
 */
func GetCallerFuncName() string {
	// 获取调用者函数的名称
	pc, _, _, _ := runtime.Caller(2)
	name := runtime.FuncForPC(pc).Name()
	names := strings.Split(name, ".")
	return names[len(names)-1]
}
